syntax = "proto3";

// Implements types for the Bufsteam demo.
package osprey;
option go_package = "./;osprey";

import "google/protobuf/timestamp.proto";

message OspreyInputEvent {
  OspreyInputEventData data = 1;
  google.protobuf.Timestamp send_time = 2;
}

message OspreyInputEventData {
  string action_name = 1;
  int64 action_id = 2;
  bytes data = 3;
  google.protobuf.Timestamp timestamp = 4;
  map<string, string> secret_data = 5;
  string encoding = 6;
}

message AtprotoLabelEffect {
  AtprotoEffectKind effect_kind = 1;
  AtprotoSubjectKind subject_kind = 2;
  AtprotoLabel label = 3;
  string comment = 4;
  optional AtprotoEmail email = 5;
  optional int64 expiration_in_hours = 6;
  repeated string rules = 7;
}

message AtprotoTagEffect {
  AtprotoEffectKind effect_kind = 1;
  AtprotoSubjectKind subject_kind = 2;
  string tag = 3;
  optional string comment = 4;
  repeated string rules = 5;
}

message AtprotoTakedownEffect {
  AtprotoEffectKind effect_kind = 1;
  AtprotoSubjectKind subject_kind = 2;
  string comment = 4;
  optional AtprotoEmail email = 5;
  repeated string rules = 6;
}

message AtprotoEmailEffect {
  AtprotoEmail email = 1;
  optional string comment = 2;
  repeated string rules = 3;
}

message AtprotoCommentEffect {
  AtprotoSubjectKind subject_kind = 1;
  string comment = 2;
  repeated string rules = 3;
}

message AtprotoEscalateEffect {
  AtprotoSubjectKind subject_kind = 1;
  optional string comment = 2;
  repeated string rules = 3;
}

message AtprotoAcknowledgeEffect {
  AtprotoSubjectKind subject_kind = 1;
  optional string comment = 2;
  repeated string rules = 3;
}

message AtprotoReportEffect {
  AtprotoSubjectKind subject_kind = 1;
  AtprotoReportKind report_kind = 2;
  string comment = 3;
  optional int64 priority_score = 4;
  repeated string rules = 5;
}

message BigQueryFlagEffect {
  AtprotoSubjectKind subject_kind = 1;
  string tag = 2;
  optional string comment = 3;
  repeated string rules = 4;
}

message ResultEvent {
  google.protobuf.Timestamp send_time = 1;
  string action_name = 2;
  int64 action_id = 3;
  string did = 4;
  string uri = 5;
  string cid = 6;
  bytes data = 7;
  repeated AtprotoLabelEffect labels = 8;
  repeated AtprotoTagEffect tags = 9;
  repeated AtprotoTakedownEffect takedowns = 10;
  repeated AtprotoEmailEffect emails = 11;
  repeated AtprotoCommentEffect comments = 12;
  repeated AtprotoEscalateEffect escalations = 13;
  repeated AtprotoAcknowledgeEffect acknowledgements = 14;
  repeated AtprotoReportEffect reports = 15;
  repeated BigQueryFlagEffect bigqueryFlags = 16; 
}

enum AtprotoSubjectKind {
  ATPROTO_SUBJECT_KIND_NONE = 0;
  ATPROTO_SUBJECT_KIND_ACTOR = 1;
  ATPROTO_SUBJECT_KIND_RECORD = 2;
}

enum AtprotoLabel {
  ATPROTO_LABEL_NONE = 0;
  ATPROTO_LABEL_SPAM = 1;
  ATPROTO_LABEL_RUDE = 2;
  ATPROTO_LABEL_PORN = 3;
  ATPROTO_LABEL_SEXUAL = 4;
  ATPROTO_LABEL_WARN = 5;
  ATPROTO_LABEL_HIDE = 6;
  ATPROTO_LABEL_NEEDS_REVIEW = 7;
  ATPROTO_LABEL_MISLEADING = 8;
}

enum AtprotoEffectKind {
  ATPROTO_EFFECT_KIND_NONE = 0;
  ATPROTO_EFFECT_KIND_ADD = 1;
  ATPROTO_EFFECT_KIND_REMOVE = 2;
}

enum AtprotoEmail {
  ATPROTO_EMAIL_NONE = 0;
  ATPROTO_EMAIL_SPAM_REPLY = 68;
  ATPROTO_EMAIL_SPAM_TAKEDOWN = 261;
  ATPROTO_EMAIL_SPAM_FAKE = 63;
  ATPROTO_EMAIL_SPAM_LABEL = 318;
  ATPROTO_EMAIL_SPAM_LABEL_24_HOURS = 302;
  ATPROTO_EMAIL_SPAM_LABEL_72_HOURS=  313;
  ATPROTO_EMAIL_ID_REQUEST = 281;
  ATPROTO_EMAIL_IMPERSONATION_LABEL = 31;
  ATPROTO_EMAIL_AUTOMOD_TAKEDOWN = 288;
  ATPROTO_EMAIL_REINSTATEMENT = 60;
  ATPROTO_EMAIL_THREAT_POST_TAKEDOWN = 26;
  ATPROTO_EMAIL_PEDO_ACCOUNT_TAKEDOWN = 43;
  ATPROTO_EMAIL_DMS_DISABLED = 295;
  ATPROTO_EMAIL_TOXIC_LIST_HIDE = 51;
}

enum AtprotoReportKind {
  ATPROTO_REPORT_KIND_NONE = 0;
  ATPROTO_REPORT_KIND_SPAM = 1;
  ATPROTO_REPORT_KIND_VIOLATION = 2;
  ATPROTO_REPORT_KIND_MISLEADING = 3;
  ATPROTO_REPORT_KIND_SEXUAL = 4;
  ATPROTO_REPORT_KIND_RUDE = 5;
  ATPROTO_REPORT_KIND_OTHER = 6;
}

message FirehoseEvent {
  string did = 1;

  google.protobuf.Timestamp timestamp = 2;
  
  EventKind kind = 3;

  Commit commit = 4;
  bytes account = 5;   // comatproto.SyncSubscribeRepos_Account as opaque bytes
  bytes identity = 6;  // comatproto.SyncSubscribeRepos_Identity as opaque bytes
}

message Commit {
  string rev = 1;
  CommitOperation operation = 2;
  string collection = 3;
  string rkey = 4;
  bytes record = 5;  // json.RawMessage as opaque bytes
  string cid = 6;
}

// Enums for the constants (optional, but recommended for type safety)
enum EventKind {
  EVENT_KIND_UNSPECIFIED = 0;
  EVENT_KIND_COMMIT = 1;
  EVENT_KIND_ACCOUNT = 2;
  EVENT_KIND_IDENTITY = 3;
}

enum CommitOperation {
  COMMIT_OPERATION_UNSPECIFIED = 0;
  COMMIT_OPERATION_CREATE = 1;
  COMMIT_OPERATION_UPDATE = 2;
  COMMIT_OPERATION_DELETE = 3;
}

message Cursor {
  int64 sequence = 1;
  bool saved_on_exit = 2;
}

message ModerationEnrichedFirehoseRecordEvent {
  string did = 1;
  google.protobuf.Timestamp timestamp = 2;
  string collection = 3;
  string rkey = 4;
  CommitOperation operation = 5;
  bytes record = 6;  // json.RawMessage as opaque bytes
  map<string, ImageDispatchResults> image_results = 7;  // map of image_cid to ImageDispatchResults

  optional bytes ozone_repo_view_detail = 8; // JSON encoded repoViewDetail from Ozone for the actor
  optional bytes did_doc = 9; // JSON encoded DID Document
  optional bytes profile_view = 10; // JSON encoded ProfileViewDetailed from AppView
  optional bytes did_audit_log = 11; // JSON encoded DID audit log

  string cid = 12;
}

message ImageDispatchResults {
  message AbyssResults {
    optional bytes raw = 1;
    optional string error = 2;
    optional bool is_abuse_match = 3;
  }

  message HiveResults {
    optional bytes raw = 1;
    optional string error = 2;
    map<string, double> classes = 3;  // map of class name to confidence score
  }

  message RetinaResults {
    optional bytes raw = 1;
    optional string error = 2;
    optional string text = 3;  // OCR text result
  }

  message PrescreenResults {
    optional bytes raw = 1;
    optional string error = 2;
    optional string decision = 3; // "nsfw", "sfw"
  }

  string cid = 1;
  optional AbyssResults abyss = 2;
  optional HiveResults hive = 3;
  optional RetinaResults retina = 4;
  optional PrescreenResults prescreen = 6;
}
